### PR TITLE
Feature/adc dma mode

### PR DESCRIPTION
**Pull request by victorandrehc**

### CODE REVIEW

File: src/modm/platform/adc/stm32/adc_impl.hpp.in [hunk 1]
```diff
@@ -257,3 +257,57 @@ modm::platform::Adc{{ id }}::acknowledgeInterruptFlags(const InterruptFlag_t fla
...
+}
+
+void
+modm::platform::Adc{{ id }}::disableDMASelection()
+{
+	ADC{{ per }}->CR2 &= ~ADC_CR2_DDS;
+}
+
+void
+modm::platform::Adc{{ id }}::setEndOfConversionSelection(EndOfConversionFlag endOfConversionFlag)
```
Comment (klsc-zeat) [inline src/modm/platform/adc/stm32/adc_impl.hpp.in]: enable/disable call maybe better?

File: src/modm/platform/adc/stm32/adc_impl.hpp.in [hunk 2]
```diff
@@ -257,3 +257,57 @@ modm::platform::Adc{{ id }}::acknowledgeInterruptFlags(const InterruptFlag_t fla
...
+}
+
+void
+modm::platform::Adc{{ id }}::disableDmaMode()
+{
+	ADC{{ per }}->CR2 &= ~ADC_CR2_DMA;
+}
+
+bool
+modm::platform::Adc{{ id }}::getAdcEnabled()
```
Comment (klsc-zeat) [inline src/modm/platform/adc/stm32/adc_impl.hpp.in:305]: You sure this doesn't exist already?

File: src/modm/platform/adc/stm32/adc_impl.hpp.in [hunk 3]
```diff
@@ -257,3 +257,57 @@ modm::platform::Adc{{ id }}::acknowledgeInterruptFlags(const InterruptFlag_t fla
 {
 	ADC{{ per }}->SR = ~flags.value;
 }
+
+uintptr_t
+modm::platform::Adc{{ id }}::getDataRegisterPointer()
```
Comment (klsc-zeat) [inline src/modm/platform/adc/stm32/adc_impl.hpp.in]: Why needed? This looks a bit like a hack.

Comment (victorandrehc) [reply src/modm/platform/adc/stm32/adc_impl.hpp.in]: we need it to pass the peripheral address to the DMA.

File: src/modm/platform/adc/stm32/adc_impl.hpp.in (hunk 2 above)
Comment (victorandrehc) [reply src/modm/platform/adc/stm32/adc_impl.hpp.in:305]: I couldn't find any other method similar to that, or that at least reads from this bit ion CR2 register.

File: src/modm/platform/adc/stm32/adc_impl.hpp.in (hunk 1 above)
Comment (victorandrehc) [reply src/modm/platform/adc/stm32/adc_impl.hpp.in]: maybe, but this bit always enables one thing: 
If it's set to 0 -> eocs flag set after the sequence
if it's set to 1-> eocs flag set after every conversion

**Review by klsc-zeat — APPROVED** [review]

File: examples/nucleo_f401re/adc_dma/AdcDma.hpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+			Adc::enableDmaMode();
+			Adc::enableDMASelection();
+			Dma::AdcChannel::start();
+		}
+	}
+
+	static void
+	setTransferErrorCallback(modm::platform::DmaBase::IrqHandler transferErrorCallback)
+	{
+		if (transferErrorCallback == nullptr) { return; }
```
Comment (klsc-zeat) [inline examples/nucleo_f401re/adc_dma/AdcDma.hpp]: does their code style allow this?
our code style would only allow:
```
if (bla)
{
    return;
}
```
or
```
if (bla)
    return;
```

Comment (klsc-zeat) [reply examples/nucleo_f401re/adc_dma/AdcDma.hpp]: lots of occurences

File: examples/nucleo_f401re/adc_dma/AdcDma.hpp [hunk 5]
```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) 2023, Zühlke (Austria) gmbh
```
Comment (klsc-zeat) [inline examples/nucleo_f401re/adc_dma/AdcDma.hpp]: correct company name is:
Zühlke Engineering (Austria) GmbH

File: examples/nucleo_f401re/adc_dma/AdcDma.hpp (hunk 4 above)
Comment (victorandrehc) [reply examples/nucleo_f401re/adc_dma/AdcDma.hpp]: they clang formater that did that. I would never do it.

File: examples/nucleo_f401re/adc_dma/main.cpp [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+// This Example uses adc to sample 2 channels in scan mode 50 times every 0.5s and store the value
+// in a buffer via dma. The adc is also set to be triggered by Timer1 compare Event, that is set in
```
Comment (klsc-zeat) [inline examples/nucleo_f401re/adc_dma/main.cpp]: via dma. <linebreak>
The adc is set to be triggered by the Timer1 compare event and than samples both channels using scan mode. The compare event is triggered for a defined number of times by configuring the Timer1 to use one pulse mode with an repetition count. <linebreak>
For more information...

File: examples/nucleo_f401re/adc_dma/main.cpp [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+	// Change MODM_LOG_LEVEL above to enable or disable these messages
+	MODM_LOG_DEBUG << "debug" << modm::endl;
+	MODM_LOG_INFO << "info" << modm::endl;
+	MODM_LOG_WARNING << "warning" << modm::endl;
+	MODM_LOG_ERROR << "error" << modm::endl;
+
+	LedD13::setOutput();
+	LedD13::reset();
+	advancedTimerConfig<Timer1>(adc_results.size() / 2 - 1);
+	Adc1Dma::initialize<SystemClock>((uintptr_t)(&adc_results[0]), adc_results.size());
```
Comment (klsc-zeat) [inline examples/nucleo_f401re/adc_dma/main.cpp:57]: maybe group the 2 Adc1Dma calls together if possible

maybe add a few comments for each step?

**Pull request merged**